File: May-LeetCoding-Challenge/11-Flood-Fill/Flood-Fill.scala
object Solution {
  def floodFill(image: Array[Array[Int]], sr: Int, sc: Int, newColor: Int): Array[Array[Int]] = {
    fill(image, sr, sc, image(sr)(sc), newColor)
    image
  }

  def fill(image: Array[Array[Int]], r: Int, c: Int, startColor: Int, newColor: Int): Unit = {
    val oldColor = image(r)(c)
    if (oldColor == startColor && oldColor != newColor) {
      image(r)(c) = newColor
      if (r - 1 >= 0) fill(image, r - 1, c, startColor, newColor)
      if (c - 1 >= 0) fill(image, r, c - 1, startColor, newColor)
      if (r + 1 < image.length) fill(image, r + 1, c, startColor, newColor)
      if (c + 1 < image(0).length) fill(image, r, c + 1, startColor, newColor)
    }
  }
}


object Solution2 {
  def floodFill(image: Array[Array[Int]], sr: Int, sc: Int, newColor: Int): Array[Array[Int]] = {
    val startColor = image(sr)(sc)

    def fill(r: Int, c: Int): Unit = {
      val oldColor = image(r)(c)
      if (oldColor == startColor && oldColor != newColor) {
        image(r)(c) = newColor
        if (r - 1 >= 0) fill(r - 1, c)
        if (c - 1 >= 0) fill(r, c - 1)
        if (r + 1 < image.length) fill(r + 1, c)
        if (c + 1 < image(0).length) fill(r, c + 1)
      }
    }

    fill(sr, sc)
    image
  }
}

File: May-LeetCoding-Challenge/11-Flood-Fill/Flood-Fill.cs
// Runtime: 248 ms, faster than 93.23% of C# online submissions for Flood Fill.
// Memory Usage: 32.7 MB, less than 73.36% of C# online submissions for Flood Fill.
public class Solution
{
    public int[][] FloodFill(int[][] image, int sr, int sc, int newColor)
    {
        //sr is starting row position, sc is column position
        if (image[sr][sc] == newColor)
        {
            return image;
        }

        DfsFill(image, sr, sc, image[sr][sc], newColor);

        return image;
    }

    public void DfsFill(int[][] image, int row, int column, int oldColor, int newColor)
    {
        if (row < 0 || row >= image.Length || column < 0 || column >= image[row].Length || image[row][column] != oldColor)
        {
            return;
        }

        image[row][column] = newColor;

        DfsFill(image, row - 1, column, oldColor, newColor);
        DfsFill(image, row + 1, column, oldColor, newColor);
        DfsFill(image, row, column - 1, oldColor, newColor);
        DfsFill(image, row, column + 1, oldColor, newColor);
    }
}